File: Q2683_「ロング・ロング・ストリング」問題/answer.rb
#!ruby
# Try CodeIQ
# author: Leonardone @ NEETSDKASU

def f(x)
    (x * Math.log10(x)).floor + 1 # 間違っていたので修正
end

def g(x)
    def h(b, t, m)
        return -1 if b > t
        i = ((t + b) / 2).floor
        j = f(i)
        if m > j
            h(i + 1, t, m)
        elsif m < j
            h(b, i - 1, m)
        else
            i
        end
    end
    h(2, 10 ** 10, x)
end

m = gets.to_i
p g(m)

File: Q2683_「ロング・ロング・ストリング」問題/answer.go
// Try CodeIQ
// author: Leonardone @ NEETSDKASU

package main

import (
    "fmt"
    "math"
)

func f(n float64) float64 {
    return 1 + math.Floor(n * math.Log10(n))
}

func g(m float64) float64 {
    b := 2.0
    t := math.Pow10(10)
    for b <= t {
        x := math.Floor((b + t) / 2)
        y := f(x)
        if y < m {
            b = x + 1
        } else if y > m {
            t = x - 1
        } else {
            return x
        }
    }
    return -1
}

func main() {
    var m float64;
    
    fmt.Scan(&m);
    
    ans := int64(g(m))
    
    fmt.Print(ans)
    
}
